ui/nuxeo-document-layout.js

```diff
@@ -0,0 +1,305 @@
+                    violation.invalid_value,
+                    p.field_name,
+                    ...Object.values(violation.constraint.parameters),
+                  ),
+                );
+              } else {
+                this._addValidationError(msg);
+              }
+              Object.values(widgets).forEach((widget) => {
+                // we can at least flag the widget `invalid`
+                widget.invalid = true;
+              });
+            } else {
+              this._addValidationError(this.i18n(violation.messageKey, violation.invalid_value, p.field_name));
+            }
+          });
+        } else {
+          this._addValidationError(this.i18n(violation.messageKey));
+        }
+      });
+    }
+
+    _buildHrefFn(tmpl) {
+      // eslint-disable-next-line no-new-func
+      return new Function(['document', 'layout'], `return \`${tmpl}\`.toLowerCase();`);
+    }
+
+    _loadLayout(document, layout, hrefFunction, hrefBase) {
+      this._resetValidationErrors();
+      if (document) {
+        if (!this.previousDocument || document.uid !== this.previousDocument.uid) {
+          this._set_href(null); // force layout restamp
+        }
+        if (!this.previousDocument || document.type === this.previousDocument.type) {
+          this._set_model({ document });
+        }
+        const base = hrefBase || pathFromUrl(this.__dataHost.importPath || this.importPath);
+        const path = [base, hrefFunction(document, layout)].join(base.slice(-1) !== '/' ? '/' : '');
+        this._set_href(path);
+      } else if (document === undefined) {
+        // XXX undefined is used to notify a cancel to inner elements
+        this._set_model({ document });
+      }
+      this.previousDocument = document;
+    }
+
+    _elementChanged() {
+      this._set_model({ document: this.document });
+      // forward document path change events
+      if (this.element) {
+        this.element.addEventListener('document-changed', (e) => {
+          this.notifyPath(e.detail.path, e.detail.value);
+        });
+      }
+      afterNextRender(this, () => {
+        // fire the `document-layout-changed` event only after flush
+        this.dispatchEvent(
+          new CustomEvent('document-layout-changed', {
+            bubbles: true,
+            composed: true,
+            detail: {
+              element: this.element,
+              layout: this.layout,
+            },
+          }),
+        );
+        this.applyAutoFocus();
+      });
+    }
+
+    _getBoundElements(property) {
+      return this.$.layout._getBoundElements(property);
+    }
+
+    _getFocusableElement(parent) {
+      if (parent && parent.shadowRoot && !parent.shadowRoot.activeElement) {
+        const nodes = Array.from(parent.shadowRoot.querySelectorAll('*')).filter((node) => {
+          const style = window.getComputedStyle(node);
+          return !node.disabled && style.display !== 'none' && style.visibility !== 'hidden';
```

Comment:
Interesting that we have this kind of logic here, similar to some helpers that we've been discussing. 🙂

Reply:
Indeed. A possible future refactoring, I'd say.